File: cs/etsm/Test/tick.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using etsm;
using System;

namespace tick
{
    public class State : etsm.State
    {
        public Action? Tick { get; private set; }

        public State(Action? enter, Action? exit, Action? tick = null)
            : base(enter, exit)
        {
            Tick = tick;
        }
    }

    public class Foo
    {
        public StateMachine<State> StateMachine { get; private set; }
        public State StateA { get; private set; }
        public State StateB { get; private set; }

        public int tickStateA = 0;
        public int tickStateB = 0;

        public Foo()
        {
            StateMachine = new StateMachine<State>();
            StateA = new State(null, null, TickA);
            StateB = new State(null, null, TickB);
        }

        private void TickA()
        {
            ++tickStateA;
        }

        private void TickB()
        {
            ++tickStateB;
        }
    }

    [TestClass]
    public class tick
    {
        [TestMethod]
        public void Test()
        {
            Foo foo = new Foo();

            Assert.IsTrue(foo.StateMachine.IsIn(null)); 
            Assert.AreEqual(0, foo.tickStateA);
            Assert.AreEqual(0, foo.tickStateB);

            foo.StateMachine.Transition(foo.StateA);
            Assert.IsTrue(foo.StateMachine.IsIn(foo.StateA));
            foo.StateMachine.CurrentState?.Tick?.Invoke();
            Assert.AreEqual(1, foo.tickStateA);
            Assert.AreEqual(0, foo.tickStateB);

            foo.StateMachine.Transition(foo.StateB);
            Assert.IsTrue(foo.StateMachine.IsIn(foo.StateB));
            foo.StateMachine.CurrentState?.Tick?.Invoke();
            Assert.AreEqual(1, foo.tickStateA);
            Assert.AreEqual(1, foo.tickStateB);

            foo.StateMachine.Transition(null);
            Assert.IsTrue(foo.StateMachine.IsIn(null));
            foo.StateMachine.CurrentState?.Tick?.Invoke();
            Assert.AreEqual(1, foo.tickStateA);
            Assert.AreEqual(1, foo.tickStateB);
        }
    }
}

File: rust/etsm/tests/tick.rs
use etsm::*;

state_machine!(
    Foo,
    Data,
    State {
        {A, None, None, Some(Data::new(Foo::tick_a))},
        {B, None, None, Some(Data::new(Foo::tick_b))}
    }
);

struct Data {
    tick: fn(&mut Foo),
}

impl Data {
    fn new(tick: fn(&mut Foo)) -> Data {
        Data { tick: tick }
    }
}

struct Foo {
    state_machine: StateMachine<State>,
    state_a_tick: i32,
    state_b_tick: i32,
}

impl Foo {
    fn new() -> Foo {
        Foo {
            state_machine: StateMachine::new(),
            state_a_tick: 0,
            state_b_tick: 0,
        }
    }

    fn tick_a(&mut self) {
        self.state_a_tick += 1;
    }

    fn tick_b(&mut self) {
        self.state_b_tick += 1;
    }

    fn tick(&mut self) {
        // tick the current state
        if let Some(statedata) = get_statedata(&self.state_machine) {
            (statedata.tick)(self);
        }
    }
}

#[test]
fn tick() {
    let mut foo = Foo::new();

    assert!(foo.state_machine.is_in(None));
    assert_eq!(foo.state_a_tick, 0);
    assert_eq!(foo.state_b_tick, 0);
    assert_eq!(foo.state_machine.current_state, None);

    transition!(&mut foo, state_machine, Some(State::A));
    assert!(foo.state_machine.is_in(Some(State::A)));
    foo.tick();
    assert_eq!(foo.state_a_tick, 1);
    assert_eq!(foo.state_b_tick, 0);

    transition!(&mut foo, state_machine, Some(State::B));
    assert!(foo.state_machine.is_in(Some(State::B)));
    foo.tick();
    assert_eq!(foo.state_a_tick, 1);
    assert_eq!(foo.state_b_tick, 1);

    transition!(&mut foo, state_machine, None);
    assert!(foo.state_machine.is_in(None));
    foo.tick();
    assert_eq!(foo.state_a_tick, 1);
    assert_eq!(foo.state_b_tick, 1);
}
